File: Day85/optimizedSolution.java

class Node {
    int data;
    Node left, right;
   Node(int item)    {
        data = item;
        left = right = null;
    }
}

class Solution {
    ArrayList<Integer> inOrder(Node root) {
        /**
         * This function returns the inorder traversal of a given binary tree
         * 
         * Approach: Recursively add the left, root and right in that order
         * 
         * Time Complexity: O(n)
         * Space Compelxity: O(h)
         **/
         
        ArrayList<Integer> out = new ArrayList<>();
        
        if(root == null) return out;
        
        out.addAll(inOrder(root.left));
        out.add(root.data);
        out.addAll(inOrder(root.right));
        
        return out;
    }
}

File: Day90/optimizedSolution.py
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None

class Solution:

    def helper(self, root, k, aux):
        
        if(root == None):
            return
            
        self.helper(root.left, k, aux)
        aux[0] += 1
        if(aux[0] == k):
            aux[1] = root.data
        self.helper(root.right, k, aux)
        
        return

        
    def kthSmallest(self, root, k): 
        '''
        This funcitons returns the kth smallest element in the given BST 
        
        Approach:
        Use inorder traversal to count upto the kth occurance
        
        Time Complexity: O(k)
        Space Complexity: O(h)
        '''

        aux = [0, -1]

        self.helper(root, k, aux)

        return aux[1]